### PR TITLE
feat: implementation icm42670 close #474

### LINKED ISSUE #474: ICM42670 IMU over I2C
### Device name

ICM-42670-P

### Device description

Ultra-low-power 6-axis IMU ICM42670 from TDK IvenSense

Can be found on the new ESP-Rust board: https://github.com/esp-rs/esp-rust-board

-  Low-Noise mode 6-axis current consumption of 0.55 mA
- Low-Power mode support for always-on experience
- Sleep Mode Current Consumption: 3.5μA
- User selectable Gyro Full-scale range (dps): ± 250/500/1000/2000
- User selectable Accelerometer Full-scale range (g): ± 2/4/8/16
- User-programmable digital filters for gyro, accel, and temp sensor
- APEX Motion Functions: Pedometer, Tilt Detection, Low-g Detection, Freefall Detection, Wake on Motion, Significant Motion Detection
- Host interface: 12.5 MHz I3CSM, 1 MHz I2C, 24 MHz SPI

### Device product page URL

https://invensense.tdk.com/products/motion-tracking/6-axis/icm-42670-p/

### Datasheet URL

https://invensense.wpenginepowered.com/wp-content/uploads/2021/07/DS-000451-ICM-42670-P-v1.0.pdf

### Can the device be purchased at any of the followings?

- [X] AliExpress
- [ ] Amazon
- [ ] ebay
- [X] Digi-Key
- [X] Mouser

### URLs to purchase the device

- https://www.mouser.de/ProductDetail/TDK-InvenSense/ICM-42670-P?qs=iLbezkQI%252BsgK8sMF5V7fQQ%3D%3D
- https://www.digikey.de/de/products/detail/tdk-invensense/ICM-42670-P/14319524?s=N4IgTCBcDaIJYGMC2AWMA2A7ABgA4gF0BfIA
- https://eu.mouser.com/ProductDetail/Espressif-Systems/ESP32-C3-DevKit-RUST-1?qs=4ASt3YYao0WvXOj9TGjU2A%3D%3D
- https://de.aliexpress.com/item/1005004418342288.html?spm=a2g0o.store_pc_allProduct.8148356.2.ce60152abtzs1k&pdp_npi=2%40dis%21USD%21US%20%2419.80%21US%20%2419.80%21%21%21%21%21%402132f6d016596820099177528eed80%2112000029115522071%21sh&gatewayAdapt=glo2deu4itemAdapt

### Other implementations

_No response_

### PR DESCRIPTION
**Pull request by janveeh**
First contribution, happy about feedback
sorry, too stupid to use git rebase

### CODE REVIEW

File: components/icm42670/icm42670.c [hunk 1]
```diff
@@ -0,0 +1,677 @@
...
+ */
+
+/**
+ * @file icm42670.c
+ *
+ * ESP-IDF driver for TDK ICM-42670-P IMU (found on ESP-RS board)
+ *
+ * Copyright (c) 2022 Jan Veeh (jan.veeh@motius.de)
+ *
+ * BSD Licensed as described in the file LICENSE
```
Comment (UncleRus) [inline components/icm42670/icm42670.c:26]: Wrong license, should be ISC

File: components/icm42670/icm42670.c [hunk 2]
```diff
@@ -0,0 +1,677 @@
...
+    vTaskDelay(pdMS_TO_TICKS(0.01)); 
+    I2C_DEV_GIVE_MUTEX(&dev->i2c_dev);
+    return ESP_OK;
+}
+
+static inline esp_err_t write_mreg_register(icm42670_t *dev, icm42670_mreg_number_t mreg_num, uint8_t reg, uint8_t value)
+{
+    bool mclk_rdy;
+    CHECK(icm42670_get_mclk_rdy(dev, &mclk_rdy));
+    if(!mclk_rdy){
```
Comment (UncleRus) [inline components/icm42670/icm42670.c:296]: Please apply clang-format (we have clang-format config file in repo root)

File: components/icm42670/icm42670.c [hunk 3]
```diff
@@ -0,0 +1,677 @@
...
+    return ESP_OK;
+}
+
+esp_err_t icm42670_set_gyro_pwr_mode(icm42670_t *dev, icm42670_gyro_pwr_mode_t pwr_mode)
+{
+    CHECK_ARG(dev);
+
+    CHECK(manipulate_register(dev, ICM42670_REG_PWR_MGMT0, ICM42670_GYRO_MODE_BITS, ICM42670_GYRO_MODE_SHIFT, pwr_mode));
+    // no register writes should be performed within the next 200us
+    vTaskDelay(pdMS_TO_TICKS(0.3));
```
Comment (UncleRus) [inline components/icm42670/icm42670.c:409]: `ets_delay_us()`

File: components/icm42670/icm42670.h [hunk 4]
```diff
@@ -0,0 +1,637 @@
...
+{
+    ICM42670_ACCEL_AVG_2X = 0b000,
+    ICM42670_ACCEL_AVG_4X = 0b001,
+    ICM42670_ACCEL_AVG_8X = 0b010,
+    ICM42670_ACCEL_AVG_16X = 0b011,
+    ICM42670_ACCEL_AVG_32X = 0b100,
+    ICM42670_ACCEL_AVG_64X = 0b101
+} icm42670_accel_avg_t;
+
+
```
Comment (UncleRus) [inline components/icm42670/icm42670.h:283]: Wow, is this advertising space?

File: components/icm42670/.eil.yml [hunk 5]
```diff
@@ -0,0 +1,23 @@
+---
+components:
+  - name: icm42670
+    description: Driver for TDK ICM-42670-P IMU (found on ESP-RS board)
+    group: imu
```
Comment (UncleRus) [inline components/icm42670/.eil.yml:5]: You need to describe this group by addinig it to `devtools/groups.yml`.
Also describe yourself in `devtools/persons.yml`.

File: README.md [hunk 6]
```diff
@@ -171,6 +171,12 @@ or [GitLab examples](https://gitlab.com/UncleRus/esp-idf-lib/tree/master/example
 | **sht4x**                | Driver for Sensirion SHT40/SHT41/SHT45 digital temperature and humidity sensor   | BSD-3   | `esp32`, `esp8266`, `esp32s2`, `esp32c3` | Yes
 | **si7021**               | Driver for Si7013/Si7020/Si7021/HTU2xD/SHT2x and compatible temperature and humidity sensors | BSD-3   | `esp32`, `esp32c3`, `esp8266`, `esp32s2`, `esp32c3` | Yes
 
+### Inertial measurement unit (IMUs)
```
Comment (UncleRus) [inline README.md:174]: Please do not edit README.md manually, use [this](https://github.com/UncleRus/esp-idf-lib/blob/master/CONTRIBUTING.md#updating-readmemd).
Also add yourself to the contributors list in `devtools/README.md.erb`

File: components/icm42670/icm42670.c [hunk 7]
```diff
@@ -0,0 +1,677 @@
...
+
+esp_err_t icm42670_set_accel_avg(icm42670_t *dev, icm42670_accel_avg_t avg)
+{
+    CHECK_ARG(dev);
+    return manipulate_register(dev, ICM42670_REG_ACCEL_CONFIG1, ICM42670_ACCEL_UI_AVG_BITS, ICM42670_ACCEL_UI_AVG_SHIFT, avg);
+}
+
+esp_err_t icm42670_config_int_pin(icm42670_t *dev, uint8_t int_pin, icm42670_int_config_t config)
+{
+    CHECK_ARG(dev && int_pin);
```
Comment (UncleRus) [inline components/icm42670/icm42670.c:548]: CHECK_ARG(dev && int_pin < 3);

**Comment by UncleRus** [comment]
Not bad, not bad.

After fixing minor bugs (I hope @trombik will review it too) and adding the driver page to the documentation, I'm merging this PR.

File: components/icm42670/icm42670.c [hunk 8]
```diff
@@ -0,0 +1,677 @@
...
+}
+
+static inline esp_err_t read_register(icm42670_t *dev, uint8_t reg, uint8_t *value)
+{
+    return i2c_dev_read_reg(&dev->i2c_dev, reg, value, 1);
+}
+
+static inline esp_err_t read_register_16(icm42670_t *dev, uint8_t upper_byte_reg, uint16_t *value)
+{
+    esp_err_t err;
```
Comment (trombik) [inline components/icm42670/icm42670.c:252]: `CHECK_ARG(dev && value)`

File: examples/icm42670/default/main/main.c [hunk 9]
```diff
@@ -0,0 +1,80 @@
+#include <stdio.h>
+#include <freertos/FreeRTOS.h>
+#include <freertos/task.h>
+#include <esp_system.h>
+#include "driver/gpio.h"
```
Comment (trombik) [inline examples/icm42670/default/main/main.c:5]: `#include <driver/gpio.h>`

**Comment by janveeh** [comment]
Thanks for your feedback!
worked on all comments, see last commit

**Pull request merged**